Reduce number of explicit event loops

ChristosT opened the pull request:
Instead of having all threads spin on "process_events" we can take advantage of the available schedulers of rxcpp. I used `observer_on_new_thread` but maybe `serialize_new_thread` has even lower overhead. See also https://github.com/ReactiveX/RxCpp/blob/main/DeveloperManual.md.

I worked on a separate executable but diff-ing the two will reveal the small updates.

ChristosT committed:
send: use main thread run loop for progress

ChristosT committed:
receive: Block on socket->Receive

recv is blocking but not busy wait so it's should be ok.
On shtuting down the socket, the thread is unblocked.

ChristosT committed:
service: remove event loop

The service's pipeline will be updated by the thread that pushes the request
(recv) and consumed on a dedicated thread. There shouldn' be a need for extra runloop.

ChristosT committed:
add average load

jspanchu committed:
Keep track of service subscriptions

jspanchu committed:
Keep track of response subscriptions

jspanchu committed:
Handle exitPromise when set from multiple places

jspanchu commented:
Thanks for your contribution! Just made the `shutdown` call cross platform and introduced some book-keeping for subscriptions.

jspanchu committed:
add error handling
- handle send/recv errors
- join receiver after shutdown but before closing socket
